### PR TITLE
Remove ephemeral databases support

### PR DESCRIPTION
**Pull request by viadanna**
This pull request removes support for ephemeral databases.

**Testing instructions**:

1. Spin up an OCIM devstack
2. Create an instance, checking that the `use_ephemeral_databases` option disappeared.
3. Spin up an AppServer for given instance.
4. Check that the AppServer makes use of external databases.

**Author notes and concerns**:

~1. Some unit tests are failing as the provided .env.test still does not have external servers configured. This will be fixed asap.~

**Reviewers**
- [ ] @itsjeyd 

**Changes included**
Modify models to remove field
Provides migration
Fix templates and serialization
Remove stale tests for ephemeral support
Fix remaining tests
Adjust README for given changes

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -423,17 +423,15 @@ Required settings:
   runs as.
 * `OPENSTACK_SANDBOX_SSH_USERNAME`: The user to run ansible playbooks as when
   provisioning the sandbox (default: `ubuntu`)
-* `INSTANCE_EPHEMERAL_DATABASES`: By default, instances use local mysql and mongo
-  databases. Set this to False to use external databases instead (default: True)
 * `DEFAULT_INSTANCE_MYSQL_URL`: The external MySQL database server to be used
-  by instances configured not to use ephemeral databases. The database server
+  by instances configured. The database server
```
[the remark this answers is not in the document]
Comment (viadanna) [reply README.md]: Nice 👍

File: README.md [hunk 2]
```diff
@@ -423,17 +423,15 @@ Required settings:
...
   will be represented as an instance of the `MySQLServer` model in the database.
   It is possible to create multiple instances of that model. This setting
   exists mainly to make it easier to add a MySQL database server in testing
   and development environments.  It is mandatory to set this setting to run the
   initial migrations.
 * `DEFAULT_INSTANCE_MONGO_URL`: The external MongoDB database server to be used
-  by instances configured not to use ephemeral databases. The database server
+  by instances. The database server
```
[the remark this answers is not in the document]
Comment (viadanna) [reply README.md]: 👍

File: README.md [hunk 3]
```diff
@@ -859,21 +856,13 @@ before the redeployment command. For example:
...
+to an instance, following these steps:
 
-1. Change the `INSTANCE_EPHEMERAL_DATABASES` setting to False. Note that this is
-   only necessary if you want instances to use persistent databases by default.
-   If you only want a specific instance to use persistent databases, simply set
-   the value of the `use_ephemeral_databases` field to `True` and save the instance
-   (cf. below).
-
-2. Set up external mysql and mongo databases, making a note of hostname
+1. Set up external mysql and mongo databases, making a note of hostname
```
[the remark this answers is not in the document]
Comment (viadanna) [reply README.md]: 👍

File: instance/migrations/0105_auto_20180706_1858.py [hunk 4]
```diff
@@ -0,0 +1,23 @@
+# -*- coding: utf-8 -*-
```
[the remark this answers is not in the document]
Comment (viadanna) [reply instance/migrations/0105_auto_20180706_1858.py]: Good idea.

File: instance/models/mixins/openedx_storage.py [hunk 5]
```diff
@@ -151,11 +151,6 @@ def get_storage_settings(self):
         """
         Get configuration_storage_settings to pass to a new AppServer
         """
-        if self.use_ephemeral_databases:
-            # Workaround for broken CMS course export/import
-            # caused by https://github.com/edx/edx-platform/pull/14552
-            return yaml.dump({"EDXAPP_IMPORT_EXPORT_BUCKET": ""}, default_flow_style=False)
```
[the remark this answers is not in the document]
Comment (viadanna) [reply instance/models/mixins/openedx_storage.py]: Yeah, isn't it? 😄

File: instance/models/openedx_appserver.py [hunk 6]
```diff
@@ -332,10 +331,7 @@ def create_lms_user_settings(self):
                     for user in self.lms_users.all()
                 ],
                 "django_groups": [],
-                # We do not require users to be created successfully if we're using
-                # non-ephemeral databases and have successfully provisioned any
-                # appservers for this instance in the past; we assume we got it
-                # right the first time and don't worry about errors.
+                # We assume we got it right the first time and don't worry about errors.
```
[the remark this answers is not in the document]
Comment (viadanna) [reply instance/models/openedx_appserver.py]: Alright.

File: instance/tests/models/test_openedx_instance.py [hunk 7]
```diff
@@ -243,9 +232,6 @@ def test_spawn_appserver(self, mocks, mock_provision):
                 getattr(instance, field_name),
                 getattr(appserver, field_name),
             )
-        self.assertEqual(appserver.configuration_database_settings, "")
-        ephemeral_settings = "EDXAPP_IMPORT_EXPORT_BUCKET: ''\n"
-        self.assertEqual(appserver.configuration_storage_settings, ephemeral_settings)
```
[the remark this answers is not in the document]
Comment (viadanna) [reply instance/tests/models/test_openedx_instance.py]: It actually does, forgot to commit this change, sorry.

File: instance/management/commands/migrate_ephemeral.py [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+# GNU Affero General Public License for more details.
+#
+# You should have received a copy of the GNU Affero General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+#
+"""
+    This management command will migrate instances using ephemeral databases
+    to external databases instead.
+
+    This is intended to deprecate ephemeral databases.
```
Comment (itsjeyd) [inline instance/management/commands/migrate_ephemeral.py]: @viadanna Thanks for turning the script into a management command!

Two things:

* Let's be even more explicit here and say something like:
    > This is a prerequisite for deprecating ephemeral databases, cf. https://github.com/open-craft/opencraft/pull/327.
* Nit: Any specific reason for indenting this text? I think it would make sense to remove the indentation to be consistent with the formatting of other module-level docstrings in this repo.

File: instance/management/commands/migrate_ephemeral.py [hunk 9]
```diff
@@ -0,0 +1,63 @@
...
+from django.core.management.base import BaseCommand
+
+from instance.models.openedx_instance import OpenEdXInstance
+
+LOG = logging.getLogger(__name__)
+
+
+class Command(BaseCommand):
+    """
+    migrate_ephemeral management command class
```
Comment (itsjeyd) [inline instance/management/commands/migrate_ephemeral.py]: @viadanna This is a bit generic. You could just repeat the contents of the module-level docstring here.

File: instance/management/commands/migrate_ephemeral.py [hunk 10]
```diff
@@ -0,0 +1,63 @@
...
+    """
+    help = (
+        'Disables `use_ephemeral_databases` on all instances and provisions AppServers when required.'
```
Comment (itsjeyd) [inline instance/management/commands/migrate_ephemeral.py]: @viadanna The command provisions new app servers for all instances that used to use ephemeral databases, so the "when required" part of this help text is a little confusing (at least to me :)). Let's go with something like:

```python
        'Disables ephemeral databases for all instances that are currently using them '
        'and provisions new app servers pointing to persistent databases.'
```

File: instance/management/commands/migrate_ephemeral.py [hunk 11]
```diff
@@ -0,0 +1,63 @@
...
+    )
+
+    def __init__(self, *args, **kwargs):
+        super(Command, self).__init__(*args, **kwargs)
+        self.options = {}
+        self.retried = {}
```
Comment (itsjeyd) [inline instance/management/commands/migrate_ephemeral.py]: @viadanna `self.options` and `self.retried` don't seem to be referenced anywhere. It should be OK to remove them?

File: instance/management/commands/migrate_ephemeral.py [hunk 12]
```diff
@@ -0,0 +1,63 @@
...
+
+    def handle(self, *args, **options):
+        """
+        Find instances using ephemeral databases and migrate to external ones
```
Comment (itsjeyd) [inline instance/management/commands/migrate_ephemeral.py]: @viadanna Nit: This is a little terse. Maybe just repeat the contents of the `help` text above in this docstring to beef it up a little:

```
        Disable ephemeral databases for all instances that are currently using them
        and provision new app servers pointing to persistent databases.
```

File: instance/management/commands/migrate_ephemeral.py [hunk 13]
```diff
@@ -0,0 +1,63 @@
...
+        """
+
+        # Identify instances that need updating
+        instances = OpenEdXInstance.objects.filter(use_ephemeral_databases=True)
+        LOG.info('Found "%d" instances using ephemeral databases', instances.count())
+
+        for instance in instances:
+            # Disable ephemeral support and spawn a new AppServer
+            LOG.info('Migrating %s', instance.name)
```
Comment (itsjeyd) [inline instance/management/commands/migrate_ephemeral.py]: @viadanna Let's be explicit about what the migration does and extend the message to:

```python
            LOG.info('Migrating %s to use persistent databases', instance.name)
```

File: instance/management/commands/migrate_ephemeral.py [hunk 14]
```diff
@@ -0,0 +1,63 @@
...
+            instance.use_ephemeral_databases = False
+            instance.save()
+            instance.spawn_appserver(mark_active_on_success=True, num_attempts=3)
+            LOG.info('Successfully migrated and provisioned a new AppServer.')
```
Comment (itsjeyd) [inline instance/management/commands/migrate_ephemeral.py]: @viadanna This message won't be close to the previous ones in the log files, so let's mention the name of the instance here as well:

```python
            LOG.info('Successfully migrated and provisioned a new app server for %s', instance.name)
```

File: instance/tests/models/test_openedx_instance.py [hunk 15]
```diff
@@ -243,9 +232,13 @@ def test_spawn_appserver(self, mocks, mock_provision):
...
+        storage_setting = {
```
Comment (itsjeyd) [inline instance/tests/models/test_openedx_instance.py]: @viadanna Typo: `storage_setting` should be `storage_settings`.

File: instance/tests/api/test_openedx_appserver.py [hunk 16]
```diff
@@ -162,9 +162,10 @@ def test_get_details(self):
         self.assertIn('log_error_entries', data)
 
     @patch_gandi
+    @patch('instance.models.openedx_instance.OpenEdXInstance.provision_rabbitmq')
     @patch('instance.models.openedx_appserver.OpenEdXAppServer.provision', return_value=True)
     @patch('instance.models.mixins.load_balanced.LoadBalancingServer.run_playbook')
-    def test_spawn_appserver(self, mock_run_playbook, mock_provision):
+    def test_spawn_appserver(self, mock_run_playbook, mock_provision, mock_rabbitmq):
```
Comment (itsjeyd) [inline instance/tests/api/test_openedx_appserver.py]: @viadanna It would be worth renaming `mock_rabbitmq` to `mock_provision_rabbitmq` (here and below) so that it's completely clear which method it is mocking.

File: instance/management/commands/migrate_ephemeral.py (hunk 11 above)
Comment (itsjeyd) [reply instance/management/commands/migrate_ephemeral.py]: No worries, thanks for removing the unused stuff :)

**Comment by viadanna** [comment]
Thanks @itsjeyd 

The builds are failing as an old issue with git submodules crept back. I'll get this to stage when that's resolved.

**Comment by UmanShahzad** [comment]
@viadanna Should be resolved now :+1:

File: instance/models/mixins/database.py [hunk 17]
```diff
@@ -221,6 +221,7 @@ def provision_mysql(self):
             _create_user(cursor, self.admin_user, self._get_mysql_pass(self.admin_user))
             _grant_privileges(cursor, "*", self.admin_user, "CREATE USER")
 
+            cursor.close()
```
Comment (itsjeyd) [inline instance/models/mixins/database.py]: @viadanna Just to make sure, is this change still necessary after switching to S3 storage for integration tests?

File: instance/tests/integration/integration_instance.py [hunk 18]
```diff
@@ -252,7 +253,7 @@ def assert_theme_provisioned(self, instance, appserver, application):
 
         self._assert_theme_favicon_in_html(instance, application, favicon_url)
 
-    @shard(1)
```
Comment (itsjeyd) [inline instance/tests/integration/integration_instance.py]: @viadanna Similar question here, do we need to redistribute integration tests across shards or could we keep the existing distribution?

File: instance/tests/integration/integration_instance.py [hunk 19]
```diff
@@ -293,19 +294,19 @@ def test_spawn_appserver(self):
...
 
     @shard(1)
+    @override_settings(INSTANCE_STORAGE_TYPE='s3')
     def test_betatest_accepted(self):
         """
         Provision an instance, spawn an AppServer and accepts the application.
         """
         OpenEdXInstanceFactory(
-            name='Integration - test_spawn_appserver',
+            name='Integration - test_betatest_accepted',
```
Comment (itsjeyd) [inline instance/tests/integration/integration_instance.py]: @viadanna Thanks for making instance names match test names -- that's a small but awesome change! :D

File: instance/models/mixins/database.py (hunk 17 above)
Comment (viadanna) [reply instance/models/mixins/database.py]: I don't think this is necessary for the current change, didn't really test without it. I left it as it seems a good practice.

File: instance/tests/integration/integration_instance.py (hunk 18 above)
Comment (viadanna) [reply instance/tests/integration/integration_instance.py]: Again, not strictly necessary, but the build time dropped from 2.5 hours to 1.5.

File: instance/tests/integration/integration_instance.py (hunk 19 above)
Comment (viadanna) [reply instance/tests/integration/integration_instance.py]: Yeah, this was necessary to prevent clashes in S3 bucket names during tests.

File: instance/tests/integration/integration_instance.py (hunk 18 above)
Comment (itsjeyd) [reply instance/tests/integration/integration_instance.py]: @viadanna Definitely fine to keep these changes, then 😄
